### PR TITLE
[WiP] Add multipart upload support

### PR DESCRIPTION
**Pull request by andremedeiros**
Does what it says on the title.

I've discovered, however, by implementing this that the code is hard to follow and extend.

If you're happy with this sort of style, I'll add some tests and we can get it merged. Otherwise, I'd like to recommend (and tackle) a cleanup to modernise it.

### CODE REVIEW

**Comment by kevinbirch** [comment]
@andremedeiros Is this PR completed?  I could really use multipart upload support for my use case.

**Comment by jubos** [comment]
Code overall looks pretty good, but just needs some tests and some cleanup of the individual commits for review.

**Comment by andremedeiros** [comment]
@kevinbirch yeah, it's missing a massive cleanup & tests, which we might or might not invest time on writing.

To be quite honest, at this point, the PR is here for anyone who wants to use it (we do multipart uploads, stitching files together, etc), and we use it every day, so it's up to you to decide how you want to proceed.

**Comment by jubos** [comment]
Thanks @andremedeiros for putting it together.  I will look into merging it into upstream, but will put it some testing around it.

**Comment by andremedeiros** [comment]
Thank you, but most of it was @felixhageloh ;-)

**Comment by kevinbirch** [comment]
@jubos Thanks for that.  Is this something you will be able to work on soon?  We would really love to have this feature working.  Let me know if there is anything that I can do/write to help facilitate.

**Comment by jubos** [comment]
@kevinbirch, if you want to take a stab at cleaning up the diffs on @andremedeiros's branch and giving them a few tests, I can review and merge them into master.

**Comment by omarkhan** [comment]
@andremedeiros I have made a PR to your branch with a simple test

**Comment by LinusU** [comment]
Any updates on this? Would love to see this merged

**Comment by andremedeiros** [comment]
Sorry about that. It's been merged on our fork. @omarkhan thank you for the contribution!

**Comment by LinusU** [comment]
Cool! There are currently merge conflicts with `jubos:master`, it would be great if you could rebase it on the latest master.

Hopefully @jubos could pull it then! :confetti_ball:

**Comment by LinusU** [comment]
Also, I saw that you included `ostruct` but I can't seem to find it being used anywhere. I might be wrong but are you sure that it should be included?

**Comment by andremedeiros** [comment]
@LinusU done and done! Thanks for catching that.

Big thanks to @felixhageloh for the bulk of the multipart work here, and to @omarkhan for adding the specs!

Let's get this merged! :tada: :balloon: :confetti_ball:

**Comment by jubos** [comment]
Code overall looks good!  If someone can squash it into one feature diff + tests, I can fast-forward it onto master and cut a point release.

**Comment by jubos** [comment]
I merged in #93 and cut a point release.  Thanks everyone for helping to get this over the finish line.
